src/anemoi/training/data/dataset.py

```diff
@@ -233,7 +280,11 @@ def __iter__(self) -> torch.Tensor:
             start = i - (self.multi_step - 1) * self.timeincrement
             end = i + (self.rollout + 1) * self.timeincrement
 
-            x = self.data[start : end : self.timeincrement]
+            if self.reader_group_size > 1:  # read only a subset of the grid
+                x = self.data[start : end : self.timeincrement, :, :, self.grid_start : self.grid_end]
```

Comment:
it's a bit confusing but yes, that's our current understanding. 

CPU memory usage is reduced since each worker now only stores its respective shard of the batch in pinned memory, rather than each worker replicating the entire batch.

If we look at one of the dataloader worker processes, we see that with sharding (left), we spend proportionally less time copying and more time decompressing compared to without sharding (right):

![Screenshot 2024-10-28 at 10 11 51](https://github.com/user-attachments/assets/53285a18-7a75-44c3-8235-175247d97190)

If we were able to decompress only once in shared memory, we would potentially see further improvements.